[hma] Submit Content API

BarrettOlson opened the pull request:
Summary
---------

Creates a new API by which a user can upload content to the system. Functions very similarly to  `/upload` 

this new endpoint will be `/submit/` and will take the following fields:
- submission_type: (always UPLOAD for now)
- content_id: (formerly   only used filename: this will correspond to s3 key (+ image prefix)
- content_type: str (always PHOTO for now)
- content_ref: encoded bytes of the image
- metadata: optional additional fields

Test Plan
---------
Used Postman and built out a UI that will be visible in my next PR (#536) to fully test upload functionality. Works just like the existing upload page except you can pick (must specify) an id for the content, instead of just using the filename.

eriksendc commented on hasher-matcher-actioner/hmalib/lambdas/api/submit.py:
```diff
@@ -0,0 +1,89 @@
+    submission_type: str  # TODO Enum
+    content_id: str
+    content_type: str  # TODO Enum
+    content_ref: t.Union[str, bytes]
```
I think another kind of submission might take a URL to the content. If we did that, maybe we could use content_ref for the URL and a submission type of, say, URL? More of a conversation / idea sharing comment.

BarrettOlson replied:
Yes, that is the plan!

eriksendc commented on hasher-matcher-actioner/hmalib/lambdas/api/submit.py:
```diff
@@ -0,0 +1,89 @@
+
+@dataclass
+class SubmitContentResponse(JSONifiable):
+    content_id_submitted: str
```
can we change this to content_id? I think in most APIs I've seen you keep a name consistent if it represents the same thing.

schatten commented on hasher-matcher-actioner/hmalib/lambdas/api/submit.py:
```diff
@@ -0,0 +1,89 @@
+    # The documentation below expects prefix to be '/submit/'
+    submit_api = bottle.Bottle()
+
+    @submit_api.post("/", apply=[jsoninator(SubmitContentRequestBody)])
```
Woohoo! `jsoninator` gets its first RequestType usage!

schatten commented on hasher-matcher-actioner/hmalib/lambdas/api/submit.py:
```diff
@@ -0,0 +1,89 @@
+                content_id_submitted=request.content_id, submit_successful=True
+            )
+
+        # Other possible submission types are not supported so just echo content_id for testing
```
Throw a 400 instead? Or a [422](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422) if you're feeling like it.

schatten commented on hasher-matcher-actioner/hmalib/lambdas/api/submit.py:
```diff
@@ -0,0 +1,89 @@
+        assert isinstance(request, SubmitContentRequestBody)
+        logger.debug(f"Content Submit Request Received {request.content_id}")
+
+        if request.submission_type == "UPLOAD":
```
Food for thought. If instead of upload being the sole submission type, we supported 'URL', we could hit two birds with one stone.

S3 uploads can be painlessly achieved on the client side using amplify [as described here](https://docs.amplify.aws/lib/storage/upload/q/platform/js#browser-uploads).

--- 

Also, b64 encoding adds processing overhead on client and on server. Perhaps multi-part forms would be more up your alley? You can't use JSON for serializing metadata, but you can upload file + metadata. Here's a [bottle example](https://stackoverflow.com/questions/15050064/how-to-upload-and-save-a-file-using-bottle-framework).

schatten reviewed (changes requested):
From a code organisation and structure perspective, this is great.

I have a few suggestions about the request payload. b64 encoding is resource consuming esp. for larger files. It will cause the object to be loaded into the javascript runtime on the clientside and prevent streaming on the server.

Would advise we use multi-part streaming or two phase submission. I have linked to examples for both.

schatten reviewed (approved):
Based on offline discussion, and [Amplify's weirdness](https://stackoverflow.com/questions/57855644/aws-amplify-breaks-binary-posts-to-aws-api-gateway), approving the status quo for using b64 to encode and upload file content!

BarrettOlson commented:
Thanks @eriksendc and thanks @schatten for accepting given the context, URL support sounds like something to prioritize so I'll work on that next (after the UI+storage) :)  

Going to merge, delta since last review: 
- Changed response to have `content_id`
- Also made it return 422 for not yet supported submission_types because I was feeling it (found small bug and fixed it in middleware.py because of it)
